### PR TITLE
Retry added to NvdMirror task

### PR DESCRIPTION
**Pull request by mehab**
**Description**

This PR adds retry logic to nvd mirror task that will help to deal with random gibberish seen while running NVD mirroring task as also raised here: https://github.com/jeremylong/vuln-tools/issues/42#issuecomment-1461860163
Closes issue: https://github.com/DependencyTrack/hyades/issues/489
Added corresponding unit tests as well

Screenshot 2023-04-13 at 22 18 50
Screenshot 2023-04-13 at 22 19 32

### CODE REVIEW

**Comment by jeremylong** [comment]
Thanks for the PR! seriously... This might take me a few days to get to though as I am sick and need to have a clear head to review the code.

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/nvd/NvdMirror.java [hunk 1]
```diff
@@ -58,12 +67,22 @@ public Future<?> doMirror(String ecosystem) {
...
+        Retry retry;
+        final RetryRegistry retryRegistry = RetryRegistry.of(RetryConfig.custom().
+                intervalFunction(ofExponentialBackoff(
+                        Duration.ofSeconds(config.retryBackoffInitialDurationSeconds()),
+                        config.retryBackoffMultiplier(), Duration.ofSeconds(config.retryMaxDuration())))
+                .maxAttempts(config.retryMaxAttempts())
+                .retryOnException(NvdApiException.class::isInstance)
+                .retryOnResult(response -> false)
+                .build());
+        retry = retryRegistry.retry("nvd-api");
```
Comment (nscuro) [inline mirror-service/src/main/java/org/hyades/vulnmirror/datasource/nvd/NvdMirror.java:85]: `retry` could be an `@ApplicationScoped` bean defined in `NvdMirrorConfiguration`.

Similar to how it's done in the vulnerability analyzer for circuit breakers:

https://github.com/DependencyTrack/hyades/blob/e90018987db42a8128b23424ed1ef04473974954/vulnerability-analyzer/src/main/java/org/hyades/processor/scanner/snyk/SnykProcessorConfiguration.java#L54-L70

File: mirror-service/pom.xml [hunk 2]
```diff
@@ -119,6 +120,32 @@
...
+        <dependency>
+            <groupId>io.github.resilience4j</groupId>
+            <artifactId>resilience4j-retry</artifactId>
+            <version>${lib.resilience4j.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.resilience4j</groupId>
+            <artifactId>resilience4j-ratelimiter</artifactId>
+            <version>${lib.resilience4j.version}</version>
+        </dependency>
```
Comment (nscuro) [inline mirror-service/pom.xml:133]: Do we need the rate limiter?

File: mirror-service/pom.xml [hunk 3]
```diff
@@ -119,6 +120,32 @@
...
+        <dependency>
+            <groupId>org.mock-server</groupId>
+            <artifactId>mockserver-netty</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mock-server</groupId>
+            <artifactId>mockserver-netty</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (nscuro) [inline mirror-service/pom.xml:148]: `mockserver-netty` is defined twice. Any reason why we need mockserver over WireMock, which is what all the other tests are using?

**Review by nscuro — CHANGES_REQUESTED** [review]

File: mirror-service/pom.xml (hunk 2 above)
Comment (jeremylong) [reply mirror-service/pom.xml:133]: I don't think you need the rate limiter here in your code. there is a rate limiter already built into the NVD API library being used.

Comment (jeremylong) [reply mirror-service/pom.xml:133]: I'm also improving some of the retry logic in the NVD API as I've hit other errors then the occasional "garbage" response from the NVD API.

File: mirror-service/pom.xml (hunk 3 above)
Comment (mehab) [reply mirror-service/pom.xml:148]: removed this dependency. Mistake of not cleaning up well

File: mirror-service/src/main/java/org/hyades/vulnmirror/datasource/nvd/NvdMirror.java [hunk 4]
```diff
@@ -85,6 +104,8 @@ void mirrorInternal() throws Exception {
             }
 
             updateState(new NvdMirrorState(apiClient.getLastModifiedRequest()));
+        } catch (Throwable e) {
+            throw new Exception(e);
```
[the remark this answers is not in the document]
Comment (mehab) [reply mirror-service/src/main/java/org/hyades/vulnmirror/datasource/nvd/NvdMirror.java]: changed this one

**Review by nscuro — APPROVED** [review]

**Pull request merged**